File: python/LineCluster.py

import Rhino.Geometry as rg
import Rhino.RhinoDoc as doc


class DetermineClusters():
    def __init__(self,crvs):
        self.crvs=crvs
        self.clusterList=[]

    @staticmethod
    def is_two_lines_intersect(line1,line2):
        #检查两条线是否相交
        tol = doc.ActiveDoc.ModelAbsoluteTolerance
        events=rg.Intersect.Intersection.CurveCurve(line1,line2,tol,0.0)
        if events:
            return True

    @staticmethod
    def del_dup_lines(cluster,lineList):
        #在线段列表中删除已经加入至簇的线
        for line in cluster:
            if line in lineList:
                lineList.remove(line)
        return lineList

    def first_sort_cluster(self,lines):
        #对线列表进行初始筛选，将与其中一条线相交的线同这条线一起放入一个簇
        cluster=[]
        cluster.append(lines[0])
        lines.pop(0)
        for i in range(len(lines)):
            if self.is_two_lines_intersect(cluster[0],lines[i]):
                cluster.append(lines[i])
        return cluster

    def sort_one_cluster(self,lines):
        #选出一个线段簇
        cluster=self.first_sort_cluster(lines)
        left_lines=self.del_dup_lines(cluster,lines)
        while True:
            temp_cluster=[]
            for left_line in left_lines:
                for line in cluster:
                    if self.is_two_lines_intersect(line,left_line):
                        temp_cluster.append(left_line)
                        break
            if len(temp_cluster)!=0:
                cluster.extend(temp_cluster)
                left_lines=self.del_dup_lines(cluster,left_lines)
            else:
                self.clusterList.append(cluster)
                break
        return left_lines
        
    def sort_all_cluster(self):
        #筛选出所有线段簇
        left_lines=self.crvs
        while True:
            left_lines=self.sort_one_cluster(left_lines)
            if len(left_lines)==0:
                break


class DelOverlapLines():
    def __init__(self,crvs):
        self.crvs=crvs
        self.exploded_lines=[]
        self.checked_lines=[]

    @staticmethod
    def is_two_lines_overlap(crv1,crv2):
        #检查两条线是否重叠
        tol = doc.ActiveDoc.ModelAbsoluteTolerance
        crv11=crv1.ToNurbsCurve()
        crv22=crv2.ToNurbsCurve()
        event=rg.Intersect.Intersection.CurveCurve(crv1,crv22,tol,tol)
        if event:
            return event[0].IsOverlap

    def merge_curves(self,short_line,long_line):
        #去除重合线，并组合曲线
        tol = doc.ActiveDoc.ModelAbsoluteTolerance
        sPt=short_line.PointAtStart
        ePt=short_line.PointAtEnd
        t1=long_line.ClosestPoint(sPt)[1]
        crvPt1=long_line.PointAt(t1)
        t2=long_line.ClosestPoint(ePt)[1]
        crvPt2=long_line.PointAt(t2)
        if crvPt1.DistanceTo(sPt) <tol and crvPt2.DistanceTo(ePt)<tol:
            return long_line
        else:
            lines=[]
            if crvPt1.DistanceTo(sPt) <tol:
                subLine=long_line.Split(t1)[0]
                if not self.is_two_lines_overlap(subLine,short_line):
                    lines.append(subLine)
                    lines.append(short_line)
                    return rg.Curve.JoinCurves(lines,tol)[0]
                else:
                    lines.append(long_line.Split(t1)[1])
                    lines.append(short_line)
                    return rg.Curve.JoinCurves(lines,tol)[0]
            else:
                subLine=long_line.Split(t2)[0]
                if not self.is_two_lines_overlap(subLine,short_line):
                    lines.append(subLine)
                    lines.append(short_line)
                    return rg.Curve.JoinCurves(lines,tol)[0]
                else:
                    lines.append(long_line.Split(t2)[1])
                    lines.append(short_line)
                    return rg.Curve.JoinCurves(lines,tol)[0]

    def merge_two_overlap_curve(self,line1,line2):
        #将两条重合线合并为一条线，并返回这条线
        tol = doc.ActiveDoc.ModelAbsoluteTolerance
        if line1.GetLength() > line2.GetLength():
            return self.merge_curves(line2,line1)
        elif line1.GetLength() == line2.GetLength():
            return self.merge_curves(line1,line2)
        else :
            return self.merge_curves(line1,line2)


    @staticmethod
    def explode_polyCrv(line):
        #接受一个值，返回一个列表
        sPt=line.PointAtStart
        ePt=line.PointAtEnd
        t_start=line.ClosestPoint(sPt)[1]
        t_end=line.ClosestPoint(ePt)[1]
        if line.GetNextDiscontinuity(rg.Continuity.C1_continuous,t_start,t_end)[0]:
        #检查是多段线还是一条线
            lines=[]
            while line.GetNextDiscontinuity(rg.Continuity.C1_continuous,t_start,t_end)[0]:
                t0=line.GetNextDiscontinuity(rg.Continuity.C1_continuous,t_start,t_end)[1]
                new_line=line.Split(t0)[0]
                line=line.Split(t0)[1]
                lines.append(new_line)
                t_start=t0
            lines.append(line)
            return lines
        else:
            lines=[]
            lines.append(line)
            return lines
        
    def explode_crvs(self,lines):
        #接收线列表，炸开一组线，返回一个列表
        if len(lines)>1:
            for line in lines:
                newLine_list=self.explode_polyCrv(line)
                for i in newLine_list:
                    self.exploded_lines.append(i)
        else:
            newLine_list=self.explode_polyCrv(lines[0])
            for i in newLine_list:
                self.exploded_lines.append(i)

        
    def clean_crvs(self,crvs):
        if len(crvs)==1:
            self.checked_lines.append(crvs[0])
        else:
            newCrvs=crvs
            try:
                for x in range(len(newCrvs)):
                    crv=newCrvs[x]
                    newCrvs.remove(crv)
                    for y in newCrvs:
                        if self.is_two_lines_overlap(crv,y):
                            newCrv=self.merge_two_overlap_curve(crv,y)
                            newCrvs.remove(y)
                            newCrvs.append(newCrv)
                            return self.clean_crvs(newCrvs)
                    self.checked_lines.append(crv)
            except IndexError:
                #二分法
                self.clean_crvs(newCrvs)
                
if __name__ =="__main__":
    dol=DelOverlapLines(crvs)
    dol.explode_crvs(crvs)
    exploded=dol.exploded_lines
    dol.clean_crvs(exploded)
    a=dol.checked_lines
    dc=DetermineClusters(a)
    dc.sort_all_cluster()
    count=dc.clusterList
    try:
        cluster=dc.clusterList[i]
    except IndexError as e:
        print(e)

File: c#/LineCluster.cs
using Rhino;
using Rhino.Commands;
using Rhino.DocObjects;
using Rhino.Geometry;
using Rhino.Geometry.Intersect;
using Rhino.Input;
using Rhino.Input.Custom;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Runtime.CompilerServices;

namespace LC
{
    public class LCCommand : Command
    {
        public LCCommand()
        {
            Instance = this;
        }
        public static LCCommand  Instance { get; private set; }
        public override string EnglishName => "DetermineClusters";

        protected override Result RunCommand(RhinoDoc doc, RunMode mode)
        {
            var go = new GetObject();
            go.SetCommandPrompt("Select curves");
            go.GeometryFilter = ObjectType.Curve;
            go.GetMultiple(1, 0);
            if (go.CommandResult() != Result.Success) { return go.CommandResult(); }

            List<Curve> curves = new List<Curve>();
            for (var i = 0; i < go.ObjectCount; i++)
            {
                Curve curve = go.Object(i).Curve();
                if (null != curve)
                {
                    curves.Add(curve);
                }
            }
            //炸开选取的线
            ExplodeCurves.ExplodeCrvs(curves);
            var explodedCrvs = ExplodeCurves.ExplodedCrvs;

            //去除重叠的线
            var cleanedCrvs=DelOverlapCrvs.CleanCrvs(explodedCrvs);

            //判断线有几簇
            DetermineClusters.SortAllCluster(cleanedCrvs);
            int x = DetermineClusters.ClusterList.Count;
            RhinoApp.WriteLine("{0} clusters found.", x.ToString());

            return Result.Success;
        }
    }
    public class JoinTwoOverlapCurves
    {
        //合并两条有重叠的曲线
        public static bool IsTwoCrvsOverlap(Curve crv1, Curve crv2)
        {
            //检查两条线是否重叠
            var tol = RhinoDoc.ActiveDoc.ModelAbsoluteTolerance;
            var crv11 = crv1.ToNurbsCurve();
            var crv22 = crv2.ToNurbsCurve();
            var events = Intersection.CurveCurve(crv11, crv22, tol, tol);
            if (events.Count != 0 && events[0].IsOverlap)
            {
                return true;
            }
            else { return false; }

        }

        public static Curve MergeCurves(Curve shortCrv, Curve longCrv)
        {
            //去除重合线，并组合曲线
            var tol = RhinoDoc.ActiveDoc.ModelAbsoluteTolerance;
            var sPt = shortCrv.PointAtStart;
            var ePt = shortCrv.PointAtEnd;
            bool result1 = longCrv.ClosestPoint(sPt, out double t1);
            var crvPt1 = longCrv.PointAt(t1);
            bool result2 = longCrv.ClosestPoint(ePt, out double t2);
            var crvPt2 = longCrv.PointAt(t2);

            if (crvPt1.DistanceTo(sPt) < tol && crvPt2.DistanceTo(ePt) < tol)
            {
                return longCrv;
            }
            else
            {
                var lines = new List<Curve>();
                if (crvPt1.DistanceTo(sPt) < tol)
                {
                    var subLine = longCrv.Split(t1)[0];
                    if (IsTwoCrvsOverlap(subLine, shortCrv) == false)
                    {
                        lines.Add(subLine);
                        lines.Add(shortCrv);
                        return Curve.JoinCurves(lines, tol)[0];
                    }
                    else
                    {
                        lines.Add(longCrv.Split(t1)[1]);
                        lines.Add(shortCrv);
                        return Curve.JoinCurves(lines, tol)[0];
                    }
                }
                else
                {
                    var subLine = longCrv.Split(t2)[0];
                    if (IsTwoCrvsOverlap(subLine, shortCrv) is false)
                    {
                        lines.Add(subLine);
                        lines.Add(shortCrv);
                        return Curve.JoinCurves(lines, tol)[0];
                    }
                    else
                    {
                        lines.Add(longCrv.Split(t2)[1]);
                        lines.Add(shortCrv);
                        return Curve.JoinCurves(lines, tol)[0];
                    }
                }
            }
        }
        public static Curve MergeTwoOverlapCurve(Curve crv1, Curve crv2)
        {
            //将两条重合线合并为一条线，并返回这条线
            var tol = RhinoDoc.ActiveDoc.ModelAbsoluteTolerance;
            if (crv1.GetLength() > crv2.GetLength())
            {
                return MergeCurves(crv2, crv1);
            }
            else if (crv1.GetLength() == crv2.GetLength())
            {
                return MergeCurves(crv1, crv2);
            }
            else
            {
                return MergeCurves(crv1, crv2);
            }
        }
    }
    public class DelOverlapCrvs
    {
        //删除一组曲线中有重叠部分的线，并将删除后的线合并
        public static List<Curve> CleanCrvs(List<Curve> crvs)
        {
            //去除一个列表中的重叠部分的线
            if (crvs.Count == 1)
            {
                return crvs;
            }
            else
            {
                int x = crvs.Count;
                var list1 = crvs.Take(x / 2).ToList();
                var list2 = crvs.Skip(x / 2).ToList();
                var crvs1 = CleanCrvs(list1);
                var crvs2= CleanCrvs(list2);

                Excute:
                    foreach (var crvi in crvs1)
                    {
                        foreach (var crvj in crvs2)
                        {
                            if (JoinTwoOverlapCurves.IsTwoCrvsOverlap(crvi, crvj))
                            {
                                var newCrv = JoinTwoOverlapCurves.MergeTwoOverlapCurve(crvi, crvj);
                                crvs2.Remove(crvj);
                                crvs2.Add(newCrv);
                                crvs1.Remove(crvi);
                            goto Excute;
                            }
                        }
                        crvs2.Add(crvi);
                    }
                return crvs2;
            }
        }
    }
    public class ExplodeCurves
    {   
        private static List<Curve> _explodedCrvs = new List<Curve>();
        public static List<Curve> ExplodedCrvs { 
            get { return _explodedCrvs; }
            set { _explodedCrvs = value; } }
        public static List<Curve> ExplodePolyCrv(Curve crv)
        {
            //接受一个值，返回一个列表
            var sPt = crv.PointAtStart;
            var ePt = crv.PointAtEnd;
            bool result1 = crv.ClosestPoint(sPt, out double t_start);
            bool result2=crv.ClosestPoint(ePt, out double t_end);
            if (crv.GetNextDiscontinuity(Continuity.C1_continuous, t_start, t_end, out double t))
            //检查是多段线还是一条线
            {
                var crvs=new List<Curve>();

                while (crv.GetNextDiscontinuity(Continuity.C1_continuous, t_start, t_end,out double t1)==true)
                {
                    double t0;
                    crv.GetNextDiscontinuity(Continuity.C1_continuous, t_start, t_end, out t0);
                    var new_line = crv.Split(t0)[0];
                    crv = crv.Split(t0)[1];
                    crvs.Add(new_line);
                    t_start = t0;
                }
                crvs.Add(crv);
                return crvs;
            }
            else {
                var lines=new List<Curve>();
                lines.Add(crv);
                return lines;
            }

        }

        public static void ExplodeCrvs(List<Curve> crvs)
        {
            //接收线列表，炸开一组线，返回一个列表
            if (crvs.Count > 1)
            {
                foreach (Curve crv in crvs)
                {
                    var newCrvList = ExplodePolyCrv(crv);
                    foreach (var newCrv in newCrvList)
                    {
                        ExplodedCrvs.Add(newCrv);
                    }
                }
            }
            else {
                var newCrvList = ExplodePolyCrv(crvs[0]);
                foreach (var newCrv in newCrvList)
                {
                    ExplodedCrvs.Add((newCrv));
                }
            }
        }
    }
    public class DetermineClusters
    {
        //判定线有多少簇
        private static Dictionary<int, List<Curve>> _clusterList = new Dictionary<int, List<Curve>>();
        public static Dictionary<int, List<Curve>> ClusterList
        {
            get { return _clusterList; }
            set { _clusterList = value; }
        }
        public static bool IsTwoCrvsIntersected(Curve crv1, Curve crv2)
        {
            //检查两条线是否相交
            var tol = Rhino.RhinoDoc.ActiveDoc.ModelAbsoluteTolerance;
            var events = Rhino.Geometry.Intersect.Intersection.CurveCurve(crv1, crv2, tol, 0.0);
            if (events.Count != 0) { return true; }
            return false;
        }
        public static List<Curve> DelDupCrvs(List<Curve> cluster, List<Curve> crvList)
        {
            //在线段列表中删除已经加入至簇的线
            foreach (var curve in cluster)
            {
                if (crvList.Contains(curve))
                    crvList.Remove(curve);
            }
            return crvList;

        }
        public static List<Curve> FirstSortCluster(List<Curve> crvs)
        {
            //对线列表进行初始筛选，将与其中一条线相交的线同这条线一起放入一个簇
            List<Curve> cluster = new List<Curve>
            {
                crvs[0]
            };
            crvs.RemoveAt(0);
            foreach (var crv in crvs)
            {
                if (IsTwoCrvsIntersected(cluster[0], crv))
                {
                    cluster.Add(crv);
                }
            }
            return cluster;
        }
        public static List<Curve> SortOneCluster(List<Curve> crvs, int i)
        {
            //选出一个线段簇
            var cluster = FirstSortCluster(crvs);
            var leftCrvs = DelDupCrvs(cluster, crvs);
            var newLeftCrvs = new List<Curve>();
            while (true)
            {
                var tempCluster = new List<Curve>();
                foreach (var leftCrv in leftCrvs)
                {
                    foreach (var crv in cluster)
                    {
                        if (IsTwoCrvsIntersected(crv, leftCrv))
                        {
                            tempCluster.Add(leftCrv);
                            break;
                        }
                    }
                    if (tempCluster.Count != 0) { break; }
                    else { continue; }
                }
                if (tempCluster.Count != 0)
                {
                    cluster.AddRange(tempCluster);
                    leftCrvs = DelDupCrvs(cluster, leftCrvs);
                }
                else
                {
                    ClusterList.Add(i, cluster);
                    newLeftCrvs = leftCrvs;
                    break;
                }
            }
            return newLeftCrvs;
        }
        public static void SortAllCluster(List<Curve> curves)
        {
            //筛选出所有线段簇
            List<Curve> leftCrvs = new List<Curve>();
            leftCrvs = curves;
            int i = 0;
            while (true)
            {

                leftCrvs = SortOneCluster(leftCrvs, i);
                if (leftCrvs.Count == 0)
                {
                    break;
                }
                else { i++; continue; }
            }
        }
    }
}
